File: go-backend-interview-prep/08_http_and_backend/07_logging_and_request_id/main.py
"""Logging and request ID -- Python equivalent of the Go example.

Python stdlib has no context propagation. We generate a request ID
per request and include it in log output and response headers.
"""

import json
import os
import threading
import time
import urllib.request
from http.server import HTTPServer, BaseHTTPRequestHandler


def generate_id() -> str:
    return os.urandom(8).hex()


def json_response(handler, status: int, data: dict) -> None:
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json")
    handler.end_headers()
    handler.wfile.write(json.dumps(data).encode())


class Handler(BaseHTTPRequestHandler):
    def do_GET(self):
        start = time.perf_counter()

        # Generate or accept request ID
        request_id = self.headers.get("X-Request-ID", "") or generate_id()

        if self.path == "/hello":
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.send_header("X-Request-ID", request_id)
            self.end_headers()
            body = json.dumps({"message": "hello", "request_id": request_id})
            self.wfile.write(body.encode())
        else:
            self.send_response(404)
            self.send_header("X-Request-ID", request_id)
            self.end_headers()
            self.wfile.write(b"not found")

        elapsed = time.perf_counter() - start
        print(
            f"  [LOG] request_id={request_id} "
            f"method={self.command} path={self.path} "
            f"duration={elapsed*1000:.1f}ms"
        )

    def log_message(self, *args):
        pass


def main() -> None:
    server = HTTPServer(("127.0.0.1", 0), Handler)
    host, port = server.server_address
    base = f"http://{host}:{port}"
    print(f"Server on {base}")
    print("  GET /hello -> greeting with request ID")
    threading.Thread(target=server.serve_forever, daemon=True).start()

    print("\n--- Demo ---")

    # Normal request (server generates ID)
    resp = urllib.request.urlopen(f"{base}/hello", timeout=5)
    data = json.loads(resp.read())
    rid = resp.headers.get("X-Request-ID", "")
    print(f"  response: {data}")
    print(f"  X-Request-ID header: {rid}")

    # Request with client-provided ID
    req = urllib.request.Request(
        f"{base}/hello",
        headers={"X-Request-ID": "client-trace-abc123"},
    )
    resp2 = urllib.request.urlopen(req, timeout=5)
    data2 = json.loads(resp2.read())
    print(f"  with client ID: {data2}")

    server.shutdown()
    print("\nDone.")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/08_http_and_backend/07_logging_and_request_id/main.go
package main

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"log"
	"net"
	"net/http"
	"time"
)

// contextKey is a private type to avoid context key collisions.
type contextKey string

const reqIDKey contextKey = "requestID"

// generateID creates a short random hex string.
func generateID() string {
	b := make([]byte, 8)
	rand.Read(b)
	return hex.EncodeToString(b)
}

// --- Middleware: request ID ---
func requestIDMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Accept incoming ID or generate new one
		id := r.Header.Get("X-Request-ID")
		if id == "" {
			id = generateID()
		}
		ctx := context.WithValue(r.Context(), reqIDKey, id)
		w.Header().Set("X-Request-ID", id)
		next.ServeHTTP(w, r.WithContext(ctx))
	})
}

// statusRecorder captures the status code written by the handler.
type statusRecorder struct {
	http.ResponseWriter
	status int
}

func (sr *statusRecorder) WriteHeader(code int) {
	sr.status = code
	sr.ResponseWriter.WriteHeader(code)
}

// --- Middleware: structured logging ---
func loggingMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		rec := &statusRecorder{ResponseWriter: w, status: 200}
		next.ServeHTTP(rec, r)

		id, _ := r.Context().Value(reqIDKey).(string)
		fmt.Printf("  [LOG] request_id=%s method=%s path=%s status=%d duration=%s\n",
			id, r.Method, r.URL.Path, rec.status, time.Since(start))
	})
}

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v)
}

func helloHandler(w http.ResponseWriter, r *http.Request) {
	id, _ := r.Context().Value(reqIDKey).(string)
	writeJSON(w, http.StatusOK, map[string]string{
		"message":    "hello",
		"request_id": id,
	})
}

func main() {
	mux := http.NewServeMux()
	mux.HandleFunc("/hello", helloHandler)

	// Chain: requestID -> logging -> mux
	handler := requestIDMiddleware(loggingMiddleware(mux))

	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		log.Fatal(err)
	}
	addr := ln.Addr().String()
	fmt.Printf("Server on http://%s\n", addr)
	fmt.Println("  GET /hello -> greeting with request ID")

	server := &http.Server{Handler: handler}

	// Self-test
	go func() {
		time.Sleep(100 * time.Millisecond)
		base := "http://" + addr
		fmt.Println("\n--- Demo ---")

		// Normal request (server generates ID)
		resp, _ := http.Get(base + "/hello")
		if resp != nil {
			var body map[string]string
			json.NewDecoder(resp.Body).Decode(&body)
			resp.Body.Close()
			fmt.Printf("  response: %v\n", body)
			fmt.Printf("  X-Request-ID header: %s\n", resp.Header.Get("X-Request-ID"))
		}

		// Request with client-provided ID
		req, _ := http.NewRequest("GET", base+"/hello", nil)
		req.Header.Set("X-Request-ID", "client-trace-abc123")
		resp2, _ := http.DefaultClient.Do(req)
		if resp2 != nil {
			var body map[string]string
			json.NewDecoder(resp2.Body).Decode(&body)
			resp2.Body.Close()
			fmt.Printf("  with client ID: %v\n", body)
		}

		fmt.Println("\nDemo done.")
		server.Close()
	}()

	if err := server.Serve(ln); err != http.ErrServerClosed {
		log.Fatal(err)
	}
}
